clip.md

```diff
@@ -0,0 +1,443 @@
+   => #fun(TOP
+   => #fun(COIN
+   => #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then call CLIP_VAT . suck CLIP_VOW KPR COIN #else . #fi
+   ~> emitKick (KICKS +Int 1) TOP TAB LOT USR KPR COIN
+   ) ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then ( CLIP_TIP +Wad ( Rad2Wad(TAB) *Wad CLIP_CHIP ) ) #else wad(0) #fi )
+   ) ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  /Ray CLIP_BUF )
+   ... </k>
+
+      <clip-kicks> KICKS => KICKS +Int 1 </clip-kicks>
+      <clip-active> CLIP_ACTIVE => CLIP_ACTIVE ListItem(KICKS +Int 1)  </clip-active>
+      <clip-sales> ... KICKS +Int 1 |-> ClipSale( ... pos: (_ =>  size(CLIP_ACTIVE)), tab: (_ => TAB), usr: (_ => USR), tic: (_ => NOW), top: (_=> ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  *Ray CLIP_BUF ))  ) </clip-sales>
+      <clip-stopped> CLIP_STOPPED </clip-stopped>
+      <clip-vow> CLIP_VOW </clip-vow>
+      <clip-buf> CLIP_BUF </clip-buf>
+      <clip-vat> CLIP_VAT </clip-vat>
+      <clip-ilk> CLIP_ILK </clip-ilk>
+      <spot-ilks> CLIP_ILK |-> SpotIlk(... pip: VALUE ) </spot-ilks>
+      <spot-par> PAR </spot-par>
+      <clip-tip> CLIP_TIP </clip-tip>
+      <clip-chip> CLIP_CHIP </clip-chip>
+      <current-time> NOW </current-time>
+
+      requires CLIP_STOPPED <ClipStop noNewKick
+      andBool TAB >Rad rad(0)
+      andBool LOT >Wad wad(0)
+      andBool USR =/=K 0:Address
+      andBool KICKS +Int 1 >Int 0
+      andBool ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  *Ray CLIP_BUF ) >Ray ray(0)
+```
+
+```k
+endmodule
+```
+
+A PRE module was introduced to avoid circular imports.
+
+```k
+module CLIP
+    import DOG
+    import PRE-CLIP
+```
+
+```k
+   syntax ClipStep ::= "upchost"
+// -------------------------
+   rule <k> Clip . upchost => . ... </k>
+      <clip-chost> _ => DUST *Rad CHOP </clip-chost>
+      <clip-ilk> CLIP_ILK </clip-ilk>
+      <vat-ilks> ... CLIP_ILK |-> Ilk( ... dust: DUST) ... </vat-ilks>
+      <dog-ilks> ... CLIP_ILK |-> Ilk( ... chop: CHOP) ... </dog-ilks>
+```
+
+
+```k
+   syntax ClipLockStep ::= "redo" Int Address
+// ------------------------------------------
+   rule <k> Clip . redo ID KPR
+   => #fun(PRICE
+   => #fun(DONE
+   => #fun(TOP_FINAL
+   => #fun(COIN
+   => ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then ( #if ( TAB >=Rad CLIP_CHOST andBool ( Wad2Ray(LOT) *Ray ( Wad2Ray(VALUE) /Ray SPOT_PAR ) ) >=Ray CLIP_CHOST ) #then call CLIP_VAT . suck CLIP_VOW KPR COIN #else . #fi ) #else . #fi )
+   ~> DONE
+   ~> emitRedo ID TOP_FINAL TAB LOT USR KPR COIN
+   ) ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) )  #then ( #if ( TAB >=Rad CLIP_CHOST andBool ( Wad2Ray(LOT) *Ray ( Wad2Ray(VALUE) /Ray SPOT_PAR ) ) >=Ray CLIP_CHOST ) #then CLIP_TIP +Rad ( Rad2Wad(TAB) *Wad CLIP_CHIP ) #else wad(0) #fi ) #else wad(0) #fi )
+   )( ( Wad2Ray(VALUE) /Ray SPOT_PAR ) *Ray CLIP_BUF )
+   )( ( (NOW -Int TIC ) >Int CLIP_TAIL ) orBool ( (PRICE /Ray TOP) <Ray CLIP_CUSP ) )
+   )( Abacus CLIP_ABACUS . price TOP (NOW -Int TIC) )
+   ... </k>
```

Comment:
I find this pretty hard to read/follow, is there a way to maybe indent it or otherwise format it to make it clearer? Or is this just the standard way to do thing in K and I need to get used to it :joy: ?

Reply:
I agree that it gets hard to follow fast but imo this method of declaring local variables is the cleanest/most understandable way of making a big function such as Clip's take readable. We can go over it in the meeting so you'll get a clearer understanding of it 🙌